File: architecture-skills/design-patterns/examples/decorator/decorator.cs
// Decorator pattern in C#.
//
// Add-ons wrap another IBeverage behind the same interface — the same
// shape as System.IO stream wrappers: new GZipStream(new FileStream(...)).

using System;

interface IBeverage
{
    string Description { get; }
    double Cost { get; }
}

class Espresso : IBeverage
{
    public string Description => "espresso";
    public double Cost => 2.00;
}

abstract class AddOn : IBeverage
{
    protected readonly IBeverage Inner;

    protected AddOn(IBeverage inner) => Inner = inner;

    public abstract string Description { get; }
    public abstract double Cost { get; }
}

class Milk : AddOn
{
    public Milk(IBeverage inner) : base(inner) {}
    public override string Description => Inner.Description + " + milk";
    public override double Cost => Inner.Cost + 0.30;
}

class Mocha : AddOn
{
    public Mocha(IBeverage inner) : base(inner) {}
    public override string Description => Inner.Description + " + mocha";
    public override double Cost => Inner.Cost + 0.50;
}

class WhippedCream : AddOn
{
    public WhippedCream(IBeverage inner) : base(inner) {}
    public override string Description => Inner.Description + " + whipped cream";
    public override double Cost => Inner.Cost + 0.40;
}

class Program
{
    static void Main()
    {
        var plain = new Espresso();
        var fancy = new WhippedCream(new Mocha(new Milk(new Espresso())));
        foreach (IBeverage drink in new IBeverage[] { plain, fancy })
        {
            Console.WriteLine($"{drink.Description}: ${drink.Cost:F2}");
        }
    }
}

File: architecture-skills/design-patterns/examples/decorator/decorator.swift
// Decorator pattern in Swift.
//
// Add-ons wrap another Beverage behind the same protocol. Note the
// difference from protocol extensions, which add behavior to *every*
// conforming type at compile time; wrapping decorates individual
// instances at runtime.

import Foundation

protocol Beverage {
    var description: String { get }
    var cost: Double { get }
}

struct Espresso: Beverage {
    let description = "espresso"
    let cost = 2.00
}

struct AddOn: Beverage {
    let inner: Beverage
    let label: String
    let surcharge: Double

    var description: String { "\(inner.description) + \(label)" }
    var cost: Double { inner.cost + surcharge }
}

func milk(_ inner: Beverage) -> Beverage {
    AddOn(inner: inner, label: "milk", surcharge: 0.30)
}

func mocha(_ inner: Beverage) -> Beverage {
    AddOn(inner: inner, label: "mocha", surcharge: 0.50)
}

func whippedCream(_ inner: Beverage) -> Beverage {
    AddOn(inner: inner, label: "whipped cream", surcharge: 0.40)
}

let plain: Beverage = Espresso()
let fancy = whippedCream(mocha(milk(Espresso())))

for drink in [plain, fancy] {
    print("\(drink.description): $\(String(format: "%.2f", drink.cost))")
}
